### ISSUE TITLE
Wrong order of DefaultValueResolver in Container::getInvoker()

### ISSUE THREAD

**Issue by kakenbok** [PHP-DI/PHP-DI#693]
The `DefaultValueResolver` currently gets executed before the `TypeHintContainerResolver`. This causes type hinted parameters with default values always to resolve to the default value. Thus, it's impossible to overload php functions with new parameters, which requires any additional parameter to default to `null` in order to prevent the PHP warning: `Declaration of ... should be compatible with ...`

```
public function call(Service $service = null) {
  // $service is always null
}

$container->set(Service::class, $service);
$container->call([$obj, 'call']);
```
**Fix1**

`Container.php:402` needs to be changed from:

```
$parameterResolver = new ResolverChain([
    ...
    new DefaultValueResolver,
    new TypeHintContainerResolver($this->delegateContainer)
]);
```

to:

```
$parameterResolver = new ResolverChain([
    ...
    new TypeHintContainerResolver($this->delegateContainer),
    new DefaultValueResolver
]);
```
**Or Fix2**

Allow for injecting:
- a custom `ResolverChain`
- or a custom `Invoker` instance or class

**Comment by kakenbok** [comment]
Fix2 already has an issue: Allow for custom parameter resolvers #553

**Comment by kakenbok** [comment]
Pull request with fix + test: #694

**Comment by mnapoli** [comment]
Closed per https://github.com/PHP-DI/PHP-DI/pull/694: this is the intended behavior.
